### PR TITLE
ARROW-1652: [JS] housekeeping, vector cleanup

### PR DESCRIPTION
**Pull request by trxcllnt**
This PR addresses the first few issues in the [JS roadmap doc](https://docs.google.com/document/d/142dek89oM2TVI2Yql106Zo8IB1Ff_9zDg_EG6jPWS0M) I sent out a week or so ago. Sorry for the big PR, the housekeeping and vector cleanup work were pretty co-dependent.

JIRA issues addressed by this PR:
[ARROW-1032](https://issues.apache.org/jira/browse/ARROW-1032) - Support custom_metadata
[ARROW-1651](https://issues.apache.org/jira/browse/ARROW-1651) - Lazy row accessor in Table
[ARROW-1652](https://issues.apache.org/jira/browse/ARROW-1652) - Separate Vector into BatchVector and CompositeVector

Tasks from the roadmap (some not in JIRA):

##### Housekeeping
1. Enable the strict-mode tsc compiler settings in the build
2. Compile mjs files for node 8.x ESModules
3. Compile ES6 UMD target with native iterators/generators

##### Vector
1. Refactor Vector types to primitive forms representing the portion of a column in a single RecordBatch
2. Add Column Vector that represents primitive Vectors across RecordBatches as an entire column
3. Refactor linear column-to-batch-index lookup in `Vector.get(i)`
4. Simplify inheritance hierarchy/generic types with Traits (e.g. Nullable, Iterable, and Typed numeric variants)

##### Table
1. Implement lazy row accessor 
2. Share API/row logic with StructVector

cc: @wesm @TheNeuralBit

### CODE REVIEW

**Comment by trxcllnt** [comment]
A few things to call out:

- Vectors are local to their record batch, and concat wraps them as a [VirtualVector](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/vector/virtual.ts)
- VirtualVector now uses a [binary search](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/vector/virtual.ts#L108) to find the vector index and offset pairs
- Table and StructVector have been refactored to implement the [Struct<TKey>](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/types.ts#L68) interface
  a. Table implements Struct<string>, StructVector implements Struct<number>
  b. Table/StructVector `get` return new [Row<TKey>](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/table/row.ts) Vector for lazy access
  c. Table/StructVector implement Iterable<Row<TKey>> (is this reasonable?)
  d. Added a `select(...columns)` method that returns a Struct<TKey> of just the selected columns
  e. Added a public columns Array to get the columns directly
- Utf8Vector composes a Vector<Uint8Array>, adds a [getCodePoints](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/utf8.ts#L34) method
- The Arrow reader [uses](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/reader/vector.ts#L36) custom [Vector constructors](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/arrow.ts#L52) that mixin [nullability/Arrow field traits](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/vector/traits.ts#L28). If the constructor argument has field metadata, the constructor will return an instance of the appropriate mixin. This removes the validity checks in `get` when Vectors aren't nullable.

File: js/src/types/vector/virtual.ts
Comment (TheNeuralBit) [inline js/src/types/vector/virtual.ts]: I guess this `_from`, `_total` is just a more efficient way to return multiple values?

File: js/src/types/table/table.ts
Comment (TheNeuralBit) [reply js/src/types/table/table.ts]: d'oh nevermind just noticed that `columns` is exposed, so no one should be tempted to use `table.col(table.key(i))`, but its still true that the `Row` can't access table columns by index.

**Review by TheNeuralBit — COMMENTED** [review]
Looks really good overall! Excited to start building on these changes

File: js/src/reader/arrow.ts
Comment (trxcllnt) [reply js/src/reader/arrow.ts]: Yeah. The bang notation forces the type checker to drop nulls, but at runtime it can still be null

File: js/src/types/table/row.ts
Comment (trxcllnt) [reply js/src/types/table/row.ts]: @TheNeuralBit yeah, I'm open to ideas here. Perhaps the best approach here is to do away with the distinction between table/struct, and make `get()` and `key()` take column indexes, and `col()` take a column name. Also fwiw when a Row's `this.table` is a StructVector, the `table.col(key)` call [access the columns Array](https://github.com/trxcllnt/arrow/blob/vector-cleanup/js/src/types/table/struct.ts#L43) by index, which (if I understand correctly) sounds like what you're looking for here?

File: js/src/types/vector/traits.ts
Comment (trxcllnt) [reply js/src/types/vector/traits.ts]: Yep, totally cool w/ me. Not super happy with `readable` as a name, but I wanted to stick to normal mixin naming conventions, and readable really intends to mixin vector metadata from the Arrow reader. The Java implementation has `FieldVector`, so I considered `mixinField` and `mixinNullable`. Of course if we want to switch from verbs to nouns, `fieldMixin` and `nullableMixin` also have a nice ring.

File: js/src/types/vector/virtual.ts
Comment (trxcllnt) [reply js/src/types/vector/virtual.ts]: Yeah, essentially to avoid the tuple allocation. We might be inhibiting the JITs ability to inline these calls (but maybe not, since they're local to the module). I checked the CC/Uglify output of the return version, and neither of them inline the calls. Will test perf of both locally to see if there's any difference.

File: js/src/types/table/row.ts
Comment (TheNeuralBit) [reply js/src/types/table/row.ts]: I think your suggestion to change `get()` and `key()` to take column indexes does apply to my other comment on the `TableVector` though. I'll respond there.

File: js/src/types/table/table.ts
Comment (TheNeuralBit) [reply js/src/types/table/table.ts]: > @TheNeuralBit yeah, I'm open to ideas here. Perhaps the best approach here is to do away with the distinction between table/struct, and make get() and key() take column indexes, and col() take a column name. Also fwiw when a Row's this.table is a StructVector, the table.col(key) call accesses the columns Array by index, which (if I understand correctly) sounds like what you're looking for here?

Yeah that is what I'm getting at here - it would be great if there was a way for `Row` to access columns by index (as well as by name).
And actually now that I think about it, `StructVector` should have column names for its children, since they're specified [as a list of Fields in the Schema](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L286). So your suggestion to remove the distinction completely makes a lot of sense.

I'd be fine with leaving this PR as-is and opening another ticket for that task.

File: js/src/types/vector/traits.ts
Comment (TheNeuralBit) [reply js/src/types/vector/traits.ts]: `fieldMixin` and `nullableMixin` sound good to me

File: js/src/types/table/table.ts
Comment (trxcllnt) [reply js/src/types/table/table.ts]: @TheNeuralBit I did that yesterday on my flight, and only just got the chance to [push it up](https://github.com/apache/arrow/pull/1273/commits/48c6ca48bb3490cd5d931420288c9d9e5b5c90bb) as you were commenting! One issue I ran into with accessing columns by name is columns aren't required to have names, so it's not reliable. Also yeah I misunderstood what you were asking about the Table, we should be able to special case for rows across columns from the same batch.

**Comment by trxcllnt** [comment]
@wesm just checking, are you waiting on me to merge this? if so, is there a PR merge script or process I should follow?

**Comment by wesm** [comment]
@trxcllnt see the README inhttps://github.com/apache/arrow/tree/master/dev. I can merge this later today once the build is green if you aren't all set up to push commits yet

**Review by wesm — APPROVED** [review]
+1. Build failing for unrelated reasons. thanks @trxcllnt and @TheNeuralBit!

**Pull request closed without merging**
